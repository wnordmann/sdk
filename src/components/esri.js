/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

import fetchJsonp from 'fetch-jsonp';

import PropTypes from 'prop-types';
import {Component} from 'react';
import {connect} from 'react-redux';

import View from 'ol/view';
import EsriJsonFormat from 'ol/format/esrijson';
import GeoJsonFormat from 'ol/format/geojson';
import Proj from 'ol/proj';

import {encodeQueryObject} from '../util';
import {updateSource} from '../actions/map';

/** @module components/esri
 * @desc Provides a component which will handle ArcGIS Rest Feature Service interaction.
 */
class EsriController extends Component {
  constructor(props) {
    super(props);
    this.esri_format = new EsriJsonFormat();
    this.geojson_format = new GeoJsonFormat();
    this.view = new View();
  }

  componentDidMount() {
    this.fetchData();
  }

  shouldComponentUpdate(nextProps) {
    if (nextProps.map.center[0] !== this.props.map.center[0] || nextProps.map.center[1] !== this.props.map.center[1] || nextProps.map.zoom !== this.props.map.zoom || nextProps.sources !== this.props.sources) {
      this.fetchData();
    }
    // This should always return false to keep
    // render() from being called.
    return false;
  }

  fetchData() {
    this.view.setCenter(Proj.fromLonLat(this.props.map.center));
    this.view.setZoom(this.props.map.zoom + 1);
    const extent = this.view.calculateExtent(this.props.mapinfo.size);
    for (let key in this.props.sources) {
      const source = this.props.sources[key];
      const bbox = JSON.stringify({
        xmin: extent[0],
        ymin: extent[1],
        xmax: extent[2],
        ymax: extent[3],
        spatialReference: {
          wkid: 102100,
        }
      });
      const params = {
        f: 'json',
        returnGeometry: true,
        spatialRel: 'esriSpatialRelIntersects',
        geometry: bbox,
        geometryType: 'esriGeometryEnvelope',
        inSR: '102100',
        outFields: '*',
        outSR: '4326',
      };
      const url = `${source.onlineResource}${source.featureLayer}/query/?${encodeQueryObject(params)}`;
      fetchJsonp(url, {timeout: this.props.timeout}).then(
        response => response.json(),
      ).then((json) => {
        const features = this.esri_format.readFeatures(json);
        this.props.dispatch(updateSource(key, {data: this.geojson_format.writeFeaturesObject(features)}));
      }).catch(function(error) {
        console.error('An error occured.', error);
      });
    }
  }

  render() {
    // never render anything.
    return false;
  }
}

EsriController.propTypes = {
  /** Request timeout in milliseconds */
  timeout: PropTypes.number,
  /** List of map sources. */
  sources: PropTypes.object,
};

EsriController.defaultProps = {
  timeout: 30000,
  sources: {},
};

function mapStateToProps(state) {
  return {
    sources: state.esri.sources,
    map: state.map,
    mapinfo: state.mapinfo,
  };
}

export default connect(mapStateToProps)(EsriController);
