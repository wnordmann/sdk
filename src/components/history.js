/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module components/history
 * @desc Track the history of the map in the hash.
 *
 */

import PropTypes from 'prop-types';
import React from 'react';
import {connect} from 'react-redux';

import {setView} from '../actions/map';
import {parseQueryString, encodeQueryObject} from '../util';

/** Attempts to parse a given value as a floating point number.
 * @param {*} value The given value to try to parse.
 * @param {number} defaultValue The fallback value to return in case value cannot by parsed as float.
 *
 * @returns {number} Value parsed as a float, or defaultValue,
 */
function proposeFloat(value, defaultValue) {
  const proposed = parseFloat(value);
  if (typeof proposed === 'number' && isFinite(proposed)) {
    return proposed;
  }
  return defaultValue;
}

export class HashHistory extends React.Component {
  /** Ensure the hash is restored if already set
   */
  componentWillMount() {
    // without this, render() will overwrite what is in the URL.
    this.onHashChange();
  }

  /** After the component has mounted,
   *  add a listener for the history changes.
   */
  componentDidMount() {
    // addEventListener is supported in IE9+
    window.addEventListener('hashchange', () => {
      this.onHashChange();
    });
  }

  shouldComponentUpdate(nextProps) {
    // this only listens for the center,zoom, and rotation
    //  to change but props.map contains a lot more data.
    return (
      this.props.map.center[0] !== nextProps.map.center[0]
      || this.props.map.center[1] !== nextProps.map.center[1]
      || this.props.map.zoom !== nextProps.map.zoom
      || this.props.map.rotation !== nextProps.map.rotation
    );
  }

  /** When the hash changes, attempt to parse a location
   *  from the URL.
   */
  onHashChange() {
    // get the hash and remove the leading '#'
    const hash = window.location.hash.substring(1);
    const parsed_hash = parseQueryString(hash);

    // pass along the now parsed hash to the delegate functions.
    this.parseLocationHash(parsed_hash);
  }

  /** Encode the location in an object appropriate
   *  for hash-encoding.
   *
   *  @returns {Object} An Object with x,y,z members.
   */
  getLocationStateObject() {
    return {
      x: this.props.map.center[0].toFixed(5),
      y: this.props.map.center[1].toFixed(5),
      z: this.props.map.zoom,
      r: this.props.map.rotation ? this.props.map.rotation : 0,
    };
  }

  /** Create an object which will be encoded in the hash.
   *
   *  This object should represent a limited amount of state
   *  which is appropriate for hash-encoding.
   *
   *  @returns {Object} A new location state object.
   */
  getStateObject() {
    return Object.assign(
      {},
      this.getLocationStateObject(),
    );
  }

  /** Parse the location from the hash.
   *
   *  @param {Object} parsedHash The object resulting from parseQueryString() of the hash.
   *
   */
  parseLocationHash(parsedHash) {
    // create copies of the current center and zoom
    const center = this.props.map.center;

    const new_center = [
      proposeFloat(parsedHash.x, center[0]),
      proposeFloat(parsedHash.y, center[1]),
    ];

    const zoom = proposeFloat(parsedHash.z, this.props.map.zoom);

    // change the view.
    this.props.setView(new_center, zoom);
  }

  /** Encode the state for the URL hash.
   *  @returns {string} The encoded url.
   */
  encodeState() {
    const st = this.getStateObject();
    return encodeQueryObject(st);
  }

  render() {
    // set the hash as appropriate.
    window.history.replaceState(null, null, window.location.pathname + `#${this.encodeState()}`);

    // this doesn't actually contribute anything to the DOM!
    return false;
  }
}

HashHistory.propTypes = {
  /** Map state */
  map: PropTypes.shape({
    /** Center of the map. */
    center: PropTypes.arrayOf(PropTypes.number),
    /** Zoom level of the map. */
    zoom: PropTypes.number,
    /** Rotation of the map in degrees. */
    rotation: PropTypes.number,
  }),
  /** setView callback function, called when the view needs to change. */
  setView: PropTypes.func,
};


HashHistory.defaultProps = {
  map: {
    center: [0, 0],
    zoom: 1,
    rotation: 0,
  },
  setView: () => {
  },
};

function mapStateToProps(state) {
  return {
    map: state.map,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setView: (center, zoom) => {
      dispatch(setView(center, zoom));
    },
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(HashHistory);
