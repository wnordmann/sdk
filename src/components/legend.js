/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

import fetch from 'isomorphic-fetch';
import PropTypes from 'prop-types';
import React from 'react';
import {connect} from 'react-redux';
import OlRender from 'ol/render';
import LineString from 'ol/geom/linestring';
import Polygon from 'ol/geom/polygon';
import Point from 'ol/geom/point';
import Feature from 'ol/feature';
import VectorLayer from 'ol/layer/vector';
import {applyStyle} from 'ol-mapbox-style';
import {jsonEquals, jsonClone, getLayerById, parseQueryString, encodeQueryObject} from '../util';
import {getFakeStyle, hydrateLayer} from './map';

/** @module components/legend
 * @desc React Component to render the legend data.
 * Create legend objects using the metadata prefixes
 *  "bnd:legend-type" and "bnd:legend-contents".
 *  Neither the Mapbox GL Spec nor the specific underlying
 *  services for vector layers have a standardized way of
 *  providing legends.  This is using the metadata provided
 *  by the layer to do so.
 *  "bnd:legend-type" can be one of "image", "html", "href"
 *  where "bnd:legend-content" would then provide the appropriate
 *  additional information.
 *  "bnd:legend-type" : "image", "bnd:legend-content" would provide
 *   the src attribute for an <img>
 *  "bnd:legend-type" : "html", "bnd:legend-content" would provide
 *   the html content for a <div>
 *  "bnd:legend-type" : "href", "bnd:legend-content" would provide
 *   the URL for html content.
 */

/** Return a div that is asynchronously populated
 *  with the content from the parameter href.
 *
 *  @param {string} href The location of the content for the div.
 *
 *  @returns {Object} A <div> element.
 */
function getRemoteLegend(href) {
  let ref = null;
  const div = (<div ref={(me) => {
    ref = me;
  }} />);

  // kick off the href fun!
  fetch(href).then(response => response.text())
    .then((html) => {
    // This is equivalent to dangerouslySetInnerHTML
      if (ref !== null) {
        ref.innerHTML = html;
      }
    });

  return div;
}

/**
 *   @param {Object} layer Mapbox GL layer.
 *
 *   @returns {(Object|null)} A <div> or <img> element, or null.
 */
export function getLegend(layer) {
  if (layer.metadata === undefined) {
    return null;
  }
  const content = layer.metadata['bnd:legend-content'];

  switch (layer.metadata['bnd:legend-type']) {
    case 'image':
      return (<img alt={layer.id} src={content} />);
    case 'html':
      // eslint-disable-next-line
      return (<div dangerouslySetInnerHTML={{ __html: content }} />);
    case 'href':
      return getRemoteLegend(content);
    default:
      // no legend provided.
      return null;
  }
}

const pointGeomCache = {};
export function getPointGeometry(size) {
  if (!pointGeomCache[size]) {
    pointGeomCache[size] = new Point([size[0] / 2, size[1] / 2]);
  }
  return pointGeomCache[size];
}

const lineGeomCache = {};
export function getLineGeometry(size) {
  if (!lineGeomCache[size]) {
    const center = [size[0] / 2, size[1] / 2];
    lineGeomCache[size] = new LineString([
      [-8 + center[0], -3 + center[1]],
      [-3 + center[0], 3 + center[1]],
      [3 + center[0], -3 + center[1]],
      [8 + center[0], 3 + center[1]]
    ]);
  }
  return lineGeomCache[size];
}

const polygonGeomCache = {};
export function getPolygonGeometry(size) {
  if (!polygonGeomCache[size]) {
    const center = [size[0] / 2, size[1] / 2];
    polygonGeomCache[size] = new Polygon([[
      [-8 + center[0], -4 + center[1]],
      [-6 + center[0], -6 + center[1]],
      [6 + center[0], -6 + center[1]],
      [8 + center[0], -4 + center[1]],
      [8 + center[0], 4 + center[1]],
      [6 + center[0], 6 + center[1]],
      [-6 + center[0], 6 + center[1]],
      [-8 + center[0], 4 + center[1]]
    ]]);
  }
  return polygonGeomCache[size];
}

/** Get the legend for a raster-type layer.
 *  Attempts to detect a WMS-type source and use GetLegendGraphic,
 *  otherwise, uses SDK specified legend metadata.
 *
 *  @param {Object} layer Mapbox GL layer object.
 *  @param {Object} layer_src Mapbox GL source object.
 *
 *  @returns {(Object[]|Object)} An array of <img> elements or a <div> element.
 */
export function getRasterLegend(layer, layer_src) {
  if (layer_src.tiles && layer_src.tiles.length > 0) {
    const tile_url = layer_src.tiles[0];
    // check to see if the url is a wms request.
    if (tile_url.toUpperCase().indexOf('SERVICE=WMS') >= 0) {
      const tile_url_parts = tile_url.split('?');
      // parse the url
      const wms_params = parseQueryString(tile_url_parts[1]);

      // normalize the keys: WMS requests are sometimes allcaps,
      //  sometimes lower cased, and sometimes (evilly so) mixed case.
      const wms_keys = Object.keys(wms_params);
      for (let i = 0, ii = wms_keys.length; i < ii; i++) {
        const key = wms_keys[i];
        const uc_key = key.toUpperCase();
        wms_params[uc_key] = wms_params[key];
      }

      // get the WMS servers URL.
      const url = tile_url_parts[0];

      // REQUEST, FORMAT, and LAYER are the three required GetLegendGraphic
      // parameters.  LAYER is populated after the optional keys are added.
      const legend_params = {
        SERVICE: 'WMS',
        REQUEST: 'GetLegendGraphic',
        FORMAT: wms_params.FORMAT,
      };

      // These are optional parameters and will not be found in
      // every WMS request. This checks for the parameter before
      // adding it in.
      // WIDTH and HEIGHT are omitted as they provide a
      // hint for the LEGEND size not the underlaying map size.
      const optional_keys = [
        'STYLE', 'FEATURETYPE', 'RULE', 'SCALE',
        'SLD', 'SLD_BODY',
        'EXCEPTIONS', 'LANGUAGE',
      ];

      for (let i = 0, ii = optional_keys.length; i < ii; i++) {
        const value = wms_params[optional_keys[i]];
        if (value !== undefined) {
          legend_params[optional_keys[i]] = value;
        }
      }

      // Build the stack of URLs for each layer. Unlike GetMap,
      // each layer needs a separate call.
      const images = [];
      const layers = wms_params.LAYERS.split(',');
      for (let i = 0, ii = layers.length; i < ii; i++) {
        const params = Object.assign({}, legend_params, {
          LAYER: layers[i],
        });
        const src = `${url}?${encodeQueryObject(params)}`;
        images.push((<img alt={layers[i]} key={layers[i]} className="sdk-legend-image" src={src} />));
      }

      return images;
    }
  }

  return getLegend(layer);
}


export class Legend extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      empty: false
    };
  }

  componentWillReceiveProps(nextProps) {
    const nextLayer = getLayerById(nextProps.layers, this.props.layerId);
    const layer = getLayerById(this.props.layers, this.props.layerId);
    return (layer !== nextLayer);
  }

  layersEqual(layer1, layer2) {
    let paintEqual = true;
    if (layer1 && layer1.paint && layer2 && layer2.paint) {
      paintEqual = jsonEquals(layer1.paint, layer2.paint);
    }
    let layoutEqual = true;
    if (layer1 && layer1.layout && layer2 && layer2.layout) {
      const layout1 = Object.assign({}, layer1.layout);
      const layout2 = Object.assign({}, layer2.layout);
      delete layout1.visibility;
      delete layout2.visibility;
      layoutEqual = jsonEquals(layout1, layout2);
    }
    return paintEqual && layoutEqual;
  }

  shouldComponentUpdate(nextProps) {
    const nextLayer = getLayerById(nextProps.layers, this.props.layerId);
    const layer = getLayerById(this.props.layers, this.props.layerId);
    const layerEqual = this.layersEqual(nextLayer, layer);
    let strokeEqual = true;
    if (this.props.strokeId) {
      const nextStrokeLayer = getLayerById(nextProps.layers, this.props.strokeId);
      const strokeLayer = getLayerById(this.props.layers, this.props.strokeId);
      strokeEqual = this.layersEqual(nextStrokeLayer, strokeLayer);
    }
    return !layerEqual || !strokeEqual;
  }

  getVectorLegend(layers, layer_src) {
    const props = this.props;
    const layer = layers[0];
    if (!layer.metadata || !layer.metadata['bnd:legend-type']) {
      const size = props.size;
      return (<canvas ref={(c) => {
        if (c !== null) {
          let vectorContext = OlRender.toContext(c.getContext('2d'), {size: size});
          const fake_style = getFakeStyle(
            props.sprite,
            layers,
            props.mapbox.baseUrl,
            props.mapbox.accessToken
          );
          const olLayer = new VectorLayer();
          const me = this;

          const onApplyStyle = () => {
            const styleFn = olLayer.getStyle();
            let geom;
            if (layer.type === 'symbol' || layer.type === 'circle') {
              geom = getPointGeometry(size);
            } else if (layer.type === 'line') {
              geom = getLineGeometry(size);
            } else if (layer.type === 'fill') {
              geom = getPolygonGeometry(size);
            }
            if (geom) {
              const properties = {};
              if (layer['source-layer']) {
                properties.layer = layer['source-layer'];
              }
              const feature = new Feature(properties);
              feature.setGeometry(geom);
              const styles = styleFn(feature);
              if (styles) {
                for (let i = 0, ii = styles.length; i < ii; ++i) {
                  vectorContext.setStyle(styles[i]);
                  vectorContext.drawGeometry(geom);
                }
              } else {
                me.setState({empty: true});
              }
            }
          };

          applyStyle(olLayer, fake_style, layer.source).then(function() {
            if (!(layer.layout && layer.layout['icon-image'])) {
              onApplyStyle();
            } else {
              olLayer.once('change', () => {
                onApplyStyle();
              });
            }
          });
        }
      }} />);
    } else {
      return getLegend(layer, layer_src);
    }
  }

  transformVectorLayer(layer) {
    let result = layer;
    if (layer.ref) {
      result = hydrateLayer(this.props.layers, layer);
    }
    if (result.filter || (result.layout && result.layout.visibility === 'none')) {
      result = jsonClone(result);
      delete result.filter;
      if (result.layout) {
        delete result.layout.visibility;
      }
    }
    return result;
  }

  /** Handles how to get the legend data based on the layer source type.
   *  @returns {Object} Call to getRasterLegend() or getLegend() to return the html element.
   */
  getLegendContents() {
    // get the layer definition
    const layer = getLayerById(this.props.layers, this.props.layerId);
    if (layer === null) {
      return null;
    }

    let source_name = layer.source;
    if (layer.ref && !layer.source) {
      const ref_layer = getLayerById(this.props.layers, layer.ref);
      source_name = ref_layer.source;
    }
    const layer_src = this.props.sources[source_name];

    switch (layer_src.type) {
      case 'raster':
        return getRasterLegend(layer, layer_src);
      // while this may seem pretty verbose,
      //  it was intentionally left here to make it
      //  easy to implement other legend handlers as
      //  is deemed appropriate.
      case 'vector':
      case 'geojson':
        const layers = [this.transformVectorLayer(layer)];
        if (this.props.strokeId) {
          const strokeLayer = getLayerById(this.props.layers, this.props.strokeId);
          if (strokeLayer !== null) {
            layers.push(this.transformVectorLayer(strokeLayer));
          }
        }
        return this.getVectorLegend(layers, layer_src);
      case 'image':
      case 'video':
      case 'canvas':
      default:
        return getLegend(layer, layer_src);
    }
  }

  render() {
    let legend_contents;
    if (this.state.empty) {
      legend_contents = this.props.emptyLegendMessage;
    } else {
      legend_contents = this.getLegendContents();
      if (legend_contents === null) {
        legend_contents = this.props.emptyLegendMessage;
      }
    }
    let className = 'sdk-legend';
    if (this.props.className) {
      className = `${className} ${this.props.className}`;
    }

    return (
      <div style={this.props.style} className={className}>
        { legend_contents }
      </div>
    );
  }
}

Legend.propTypes = {
  /** The id of the layer for which this legend is meant. */
  layerId: PropTypes.string.isRequired,
  /**
   * Id of the layer that should serve as the stroke of a polygon legend swatch.
   * Only useful for vector legends that have a separate layer for fill and stroke.
   */
  strokeId: PropTypes.string,
  /** List of layers from the store. */
  layers: PropTypes.arrayOf(PropTypes.object),
  /** List of layer sources. */
  sources: PropTypes.shape({
    /** Source to associate with the layer. */
    source: PropTypes.string,
  }),
  /** Mapbox specific configuration. */
  mapbox: PropTypes.shape({
    /** Base url to use when substituting mapbox:// type urls. */
    baseUrl: PropTypes.string,
    /** Access token of the mapbox account to use. */
    accessToken: PropTypes.string,
  }),
  /** Sprite sheet url. */
  sprite: PropTypes.string,
  /** If legend is empty, show this message. */
  emptyLegendMessage: PropTypes.string,
  /** Size of the legend, only used for vector legends. */
  size: PropTypes.arrayOf(PropTypes.number),
  /** Style config object. */
  style: PropTypes.object,
  /** Css classname to apply. */
  className: PropTypes.string,
};

Legend.defaultProps = {
  size: [50, 50],
  mapbox: {
    baseUrl: '',
    accessToken: '',
  },
  layers: [],
  sources: {},
  emptyLegendMessage: undefined,
};

function mapStateToProps(state) {
  return {
    sprite: state.map.sprite,
    layers: state.map.layers,
    sources: state.map.sources,
    mapbox: state.mapbox,
  };
}

export default connect(mapStateToProps)(Legend);
