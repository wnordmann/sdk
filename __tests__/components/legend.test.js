/* global it, describe, expect, afterEach, beforeEach */

import React from 'react';
import {mount, configure} from 'enzyme';
import nock from 'nock';
import  Adapter from 'enzyme-adapter-react-16';

import {createStore, combineReducers} from 'redux';
import MapReducer from '../../src/reducers/map';

import SdkLegend, {Legend} from '../../src/components/legend';
import {getLegend, getPointGeometry, getLineGeometry, getPolygonGeometry} from '../../src/components/legend';

configure({adapter: new Adapter()});

describe('test the Legend component', () => {
  let store = null;
  const TEST_HTML = '<span>html legend</span>';
  const TEST_IMAGE = './spoof-legend.png';
  const TEST_HREF = '/spoof-legend.html';

  beforeEach(() => {
    store = createStore(combineReducers({
      map: MapReducer,
    }), {
      map: {
        version: 8,
        sources: {
          osm: {
            type: 'raster',
            tileSize: 256,
            tiles: [
              'https://a.tile.openstreetmap.org/{z}/{x}/{y}.png',
              'https://b.tile.openstreetmap.org/{z}/{x}/{y}.png',
              'https://c.tile.openstreetmap.org/{z}/{x}/{y}.png',
            ],
          },
          wms: {
            type: 'raster',
            tiles: [
              '/wms?SERVICE=WMS&LAYERS=x,y,z&FORMAT=image/png&EXCEPTIONS=image/png&BBOX={bbox-epsg-3857}',
            ],
          },
          other: {
            type: 'geojson',
            data: { },
          },
          unknown: {
            type: 'unknown',
          },
        },
        layers: [
          {
            id: 'osm',
            source: 'osm',
            metadata: {
              'bnd:legend-type': 'html',
              'bnd:legend-content': TEST_HTML,
            },
          },
          {
            id: 'wms-test',
            source: 'wms',
          },
          {
            id: 'html-test',
            source: 'other',
            metadata: {
              'bnd:legend-type': 'html',
              'bnd:legend-content': TEST_HTML,
            },
          },
          {
            id: 'href-test',
            source: 'other',
            metadata: {
              'bnd:legend-type': 'href',
              'bnd:legend-content': `http://example.com${TEST_HREF}`,
            },
          },
          {
            id: 'image-test',
            ref: 'href-test',
            metadata: {
              'bnd:legend-type': 'image',
              'bnd:legend-content': TEST_IMAGE,
            },
          },
          {
            id: 'null-test',
            source: 'other',
          },
          {
            id: 'bad-type-test',
            source: 'other',
            metadata: {
              'bnd:legend-type': 'bad-type',
            },
          }, {
            id: 'vector-point-test',
            source: 'other',
            type: 'circle',
            paint: {
              'circle-radius': 5,
              'circle-color': '#756bb1',
              'circle-stroke-color': '#756bb1',
            },
          }, {
            id: 'vector-symbol-test',
            source: 'other',
            type: 'symbol',
            layout: {
              'icon-image': 'duck',
            },
          }, {
            id: 'vector-line-test',
            source: 'other',
            type: 'line',
            layout: {
              'line-join': 'round',
              'line-cap': 'round'
            },
            'paint': {
              'line-color': '#888',
              'line-width': 8
            }
          }, {
            id: 'vector-polygon-test',
            filter: ['==', 'class', 'motorway_link'],
            'source-layer': 'foobar',
            source: 'other',
            type: 'fill',
            paint: {
              'fill-color': '#088',
              'fill-opacity': 0.8
            },
          }, {
            id: 'unknown-layer',
            source: 'unknown',
            metadata: {
              'bnd:legend-type': 'image',
              'bnd:legend-content': TEST_IMAGE,
            },
          },
        ],
      },
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it('should render the legend without error', () => {
    mount(<SdkLegend layerId="wms-test" store={store} />);
  });

  it('should render empty text', () => {
    const wrapper = mount(<Legend emptyLegendMessage='Empty Layer' layerId="wms-test" store={store} />);
    wrapper.instance().setState({empty: true});
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render a wms legend', () => {
    const wrapper = mount(<SdkLegend layerId="wms-test" store={store} />);

    // check for all the layers
    ['x', 'y', 'z'].forEach((layer) => {
      const test_src = `/wms?SERVICE=WMS&REQUEST=GetLegendGraphic&FORMAT=image%2Fpng&EXCEPTIONS=image%2Fpng&LAYER=${layer}`;
      const expected_image = (
        <img alt={layer} key={layer} className="sdk-legend-image" src={test_src} />
      );
      expect(wrapper.contains(expected_image)).toBe(true);
    });
  });

  it('should render an html legend', () => {
    const wrapper = mount(<SdkLegend layerId="html-test" store={store} />);
    expect(wrapper.html().indexOf(TEST_HTML)).toBeGreaterThan(-1);
  });

  it('should render an image legend', () => {
    const wrapper = mount(<SdkLegend layerId="image-test" store={store} />);
    expect(wrapper.html().indexOf(TEST_IMAGE)).toBeGreaterThan(-1);
  });

  it('should render an image legend if unknown source type', () => {
    const wrapper = mount(<SdkLegend layerId="unknown-layer" store={store} />);
    expect(wrapper.html().indexOf(TEST_IMAGE)).toBeGreaterThan(-1);
  });

  it('should render an href legend', (done) => {
    nock('http://example.com/')
      .get(TEST_HREF)
      .reply(200, TEST_HTML);

    const wrapper = mount(<SdkLegend layerId="href-test" store={store} />);

    // this gives the nock-fetch a short amount of time
    // to render the HTML content.
    setTimeout(() => {
      expect(wrapper.html().indexOf(TEST_HTML)).toBeGreaterThan(-1);
      done();
    }, 300);
  });

  it('should test the null state', () => {
    mount(<SdkLegend layerId="null-test" store={store} />);
  });

  it('should render an empty legend when id does not exist', () => {
    const wrapper = mount(<SdkLegend layerId="does-not-exist" store={store} />);
    expect(wrapper.find('.sdk-legend').length).toBe(1);
  });

  it('should fallback from raster to vector', () => {
    const wrapper = mount(<SdkLegend layerId="osm" store={store} />);
    expect(wrapper.html().indexOf(TEST_HTML)).toBeGreaterThan(-1);
  });

  it('should return an empty legend for an invalid legend type', () => {
    const wrapper = mount(<SdkLegend layerId="bad-type-test" store={store} />);
    expect(wrapper.find('.sdk-legend').length).toBe(1);
  });

  it('should allow for custom className', () => {
    const wrapper = mount(<SdkLegend layerId="bad-type-test" className='foo' store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should allow for vector (point) legend', () => {
    const wrapper = mount(<SdkLegend layerId="vector-point-test" store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should allow for vector (symbol) legend', () => {
    const wrapper = mount(<SdkLegend layerId="vector-symbol-test" store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should allow for vector (line) legend', () => {
    const wrapper = mount(<SdkLegend layerId="vector-line-test" store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should allow for vector (polygon) legend', () => {
    const wrapper = mount(<SdkLegend layerId="vector-polygon-test" store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should allow for custom size', () => {
    const wrapper = mount(<SdkLegend size={[100, 100]} layerId="vector-polygon-test" store={store} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should create the correct polygon geometry', () => {
    const poly = getPolygonGeometry([50, 50]);
    expect(poly.getCoordinates()).toEqual([[[17, 21], [19, 19], [31, 19], [33, 21], [33, 29], [31, 31], [19, 31], [17, 29]]]);
    const poly2 = getPolygonGeometry([50, 50]);
    expect(poly).toBe(poly2); // taken from cache
  });

  it('should create the correct point geometry', () => {
    const point = getPointGeometry([50, 50]);
    expect(point.getCoordinates()).toEqual([25, 25]);
    const point2 = getPointGeometry([50, 50]);
    expect(point).toBe(point2); // taken from cache
  });

  it('should create the correct line geometry', () => {
    const line = getLineGeometry([50, 50]);
    expect(line.getCoordinates()).toEqual([[17, 22], [22, 28], [28, 22], [33, 28]]);
    const line2 = getLineGeometry([50, 50]);
    expect(line).toBe(line2); // taken from cache
  });

  it('getLegend should return null if no metadata', () => {
    const layer = {};
    let legend = getLegend(layer);
    expect(legend).toEqual(null);
  });

  it('componentShouldReceiveProps should work correctly', () => {
    const state = store.getState().map;
    const wrapper = mount(<Legend layerId="vector-point-test" layers={state.layers} sources={state.sources} />);
    let layer;
    const layers = store.getState().map.layers;
    for (let i = 0, ii = layers.length; i < ii; ++i) {
      if (layers[i].id === 'vector-point-test') {
        layer = layers[i];
        break;
      }
    }
    const legend = wrapper.instance();
    let result = legend.componentWillReceiveProps({layers: [layer]});
    expect(result).toBe(false); // no need to update
    const newLayer = {id: 'vector-point-test'};
    result = legend.componentWillReceiveProps({layers: [newLayer]});
    expect(result).toBe(true); // layer changed so update
  });

  it('shouldComponentUpdate should work correctly', () => {
    const state = store.getState().map;
    const wrapper = mount(<Legend layerId="vector-polygon-test" strokeId="vector-line-test" layers={state.layers} sources={state.sources} />);
    let layer, stroke;
    const layers = store.getState().map.layers;
    for (let i = 0, ii = layers.length; i < ii; ++i) {
      if (layers[i].id === 'vector-polygon-test') {
        layer = layers[i];
      }
      if (layers[i].id === 'vector-line-test') {
        stroke = layers[i];
      }
    }
    const legend = wrapper.instance();
    let result = legend.shouldComponentUpdate({layers: [layer, stroke]});
    expect(result).toBe(false); // no need to update
    const newLayer = {id: 'vector-polygon-test', paint: {}};
    result = legend.shouldComponentUpdate({layers: [newLayer, stroke]});
    expect(result).toBe(true); // layer changed so update
    const newStroke = {id: 'vector-line-test', paint: {}};
    result = legend.shouldComponentUpdate({layers: [layer, newStroke]});
    expect(result).toBe(true); // stroke layer changed so update
  });

});
