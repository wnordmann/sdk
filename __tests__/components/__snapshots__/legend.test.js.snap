// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the Legend component should allow for custom className 1`] = `"<div class=\\"sdk-legend foo\\"></div>"`;

exports[`test the Legend component should allow for custom size 1`] = `"<div class=\\"sdk-legend\\"><canvas width=\\"100\\" height=\\"100\\" style=\\"width: 100px; height: 100px;\\"></canvas></div>"`;

exports[`test the Legend component should allow for vector (line) legend 1`] = `"<div class=\\"sdk-legend\\"><canvas width=\\"50\\" height=\\"50\\" style=\\"width: 50px; height: 50px;\\"></canvas></div>"`;

exports[`test the Legend component should allow for vector (point) legend 1`] = `"<div class=\\"sdk-legend\\"><canvas width=\\"50\\" height=\\"50\\" style=\\"width: 50px; height: 50px;\\"></canvas></div>"`;

exports[`test the Legend component should allow for vector (polygon) legend 1`] = `"<div class=\\"sdk-legend\\"><canvas width=\\"50\\" height=\\"50\\" style=\\"width: 50px; height: 50px;\\"></canvas></div>"`;

exports[`test the Legend component should allow for vector (symbol) legend 1`] = `"<div class=\\"sdk-legend\\"><canvas width=\\"50\\" height=\\"50\\" style=\\"width: 50px; height: 50px;\\"></canvas></div>"`;

exports[`test the Legend component should render empty text 1`] = `"<div class=\\"sdk-legend\\">Empty Layer</div>"`;
